### PR TITLE
Refactoring of export import functionality regarding new project package

### LINKED ISSUE #961: Add Library Exception
Adding a library after having deleted two libraries and added two.
The exception occurred after the second library was added.
The added library was the same as those deleted.

[webgme-client-dump (3).json.txt](https://github.com/webgme/webgme/files/228505/webgme-client-dump.3.json.txt)

### LINKED ISSUE #966: The displayed library guids are all 00000000-0000...
https://editor.dev.webgme.org/?project=demo%2BAPImported&branch=master&node=root&visualizer=ModelEditor&tab=0

### LINKED ISSUE #968: Containment and pointer rules do not propagate properly from mixins
Given the following scenario:
- there is type A and type B, both are descendants of FCO
- there is a type C, that can contain both A and B type elements
- there are also type subA which is an instance of A
- finally we also have a type mixB that is an instance of FCO but has B as mixin

In this setup, all the instances of A, B, subA, and mixB should be valid child of any instance of C. Right now, the mixB instances would be classified as invalid children of C.

The same stands for possible pointer targets.

### PR DESCRIPTION
**Pull request by kecso**
Fixes #961 
Fixes #966 
Fixes #968 

Functions for dealing with the raw json part of the new project export package have been added to storage utilities.

Needs to be squashed and cherry-picked to 1.7.x so it can be part of the 1.7.2 release as well.

### CODE REVIEW

File: src/common/storage/util.js [hunk 1]
```diff
@@ -56,6 +199,76 @@ define(['common/storage/constants', 'common/util/jsonPatcher'], function (CONSTA
             return !!(coreObj.oldHash && coreObj.newHash && coreObj.oldData && coreObj.newData);
         },
         getChangedNodes: jsonPatcher.getChangedNodes,
-        applyPatch: jsonPatcher.apply
+        applyPatch: jsonPatcher.apply,
+
+        /**
+         * Extracts a serializable json representation of a project tree.
+         * To specify starting point set one of the four options. If more than one is set the order of precedence is:
+         * rootHash, commitHash, tagName and branchName.
```
Comment (pmeijer) [inline src/common/storage/util.js]: You mentioned the different order of look up and implemented it too - please update it here too.

File: src/common/storage/util.js [hunk 2]
```diff
@@ -56,6 +199,76 @@ define(['common/storage/constants', 'common/util/jsonPatcher'], function (CONSTA
...
+         * @param {string} [parameters.commitHash] - The tree associated with the commitHash.
+         * @param {string} [parameters.tagName] - The tree at the given tag.
+         * @param {string} [parameters.branchName] - The tree at the given branch.
+         * @param {function} callback
+         */
+        getProjectJson: function (project, parameters, callback) {
+            var deferred = Q.defer(),
+                rawJson;
+
+            _getRootHash(project, parameters)
```
Comment (pmeijer) [inline src/common/storage/util.js]: `parameters || {}` - that way it won't be type error but rather your descriptive exception.

File: src/common/storage/util.js [hunk 3]
```diff
@@ -56,6 +199,76 @@ define(['common/storage/constants', 'common/util/jsonPatcher'], function (CONSTA
...
+
+            return deferred.promise.nodeify(callback);
+        },
+
+        /**
+         * Inserts a serialized project tree into the storage and associates it with a commitHash.
+         *
+         * @param {ProjectInterface} project
+         * @param {object} [options]
+         * @param {string} [options.commitMessage='information about the insertion']
```
Comment (pmeijer) [inline src/common/storage/util.js]: [options.commitMessage=%defaultCommitMessage%]

File: src/server/worker/workerrequests.js [hunk 4]
```diff
@@ -16,9 +16,9 @@ var Core = requireJS('common/core/coreQ'),
     constraint = requireJS('common/core/users/constraintchecker'),
```
Comment (pmeijer) [inline src/server/worker/workerrequests.js]: Out of curiosity are all of the additions consequences of the refactor or fixes/improvements?

File: test/common/core/librarycore.spec.js [hunk 5]
```diff
@@ -9,7 +9,7 @@ describe('Library core ', function () {
 
     var gmeConfig,
         logger,
-        Q,
+            Q,
```
Comment (pmeijer) [inline test/common/core/librarycore.spec.js]: indentation

File: src/server/worker/workerrequests.js (hunk 4 above)
Comment (kecso) [inline src/server/worker/workerrequests.js]: both. The majority of the code change is refactoring or consequences of it, but it contains small fixes as well. If anything would fall into enhancement category, that should be closely related to the refactor.

**Pull request merged**